### PR TITLE
Immutable update nested props then optionChanged

### PR DESCRIPTION
**Pull request by MikeVitik**

### CODE REVIEW

**Review by pavelgruba — DISMISSED** [review]

File: js/renovation/component_wrapper/component.ts [hunk 1]
```diff
@@ -239,6 +241,7 @@ export default class ComponentWrapper extends DOMComponent {
 
   _init() {
     super._init();
+    this._props = { ...this.option() }; //extend(true, {}, this.option());
```
Comment (vconst) [inline js/renovation/component_wrapper/component.ts]: Is this comment necessary?

File: js/renovation/ui/grids/data_grid/__tests__/data_grid.test.tsx [hunk 2]
```diff
@@ -111,6 +116,13 @@ describe('DataGrid', () => {
       expect(instance.element()).toBe(component.widgetElementRef.current);
     });
 
+    // QUnit.test('The onOptionChanged event should be called once when changing column option'
```
Comment (vconst) [inline js/renovation/ui/grids/data_grid/__tests__/data_grid.test.tsx:120]: Is this comment necessary?

File: js/renovation/component_wrapper/utils.ts [hunk 3]
```diff
@@ -29,3 +30,31 @@ export const removeDifferentElements = ($children, $newChildren): void => {
...
+  const currentPropsValue = option[name];
+  const result = props;
+  if (name !== fullName) {
+    if (Array.isArray(currentPropsValue)) {
+      const newArray = [...currentPropsValue];
+      result[name] = newArray;
+      const matchIndex = /\[\s*(\d+)\s*\]/g.exec(fullName);
+      if (matchIndex) {
+        const index = parseInt(matchIndex[1], 10);
+        if (isObject(newArray[index])) {
```
Comment (vconst) [inline js/renovation/component_wrapper/utils.ts]: Use isPlainObject here.

File: js/renovation/ui/grids/data_grid/data_grid.tsx [hunk 4]
```diff
@@ -455,14 +460,20 @@ export class DataGrid extends JSXComponent(DataGridProps) {
   // #endregion
 
   @Effect() updateOptions(): void {
-    if (this.instance && this.prevProps) {
+    if (this.instance && this.prevProps && !this.isUpdate) {
       const updatedOptions = getUpdatedOptions(this.prevProps, this.props);
+      updatedOptions.forEach(({ path }) => { this.changedOptions.push(path); });
```
Comment (vconst) [inline js/renovation/ui/grids/data_grid/data_grid.tsx]: Why is changedOptions modified here? It is not used in another places.

File: js/renovation/ui/grids/data_grid/data_grid.tsx [hunk 5]
```diff
@@ -480,40 +491,50 @@ export class DataGrid extends JSXComponent(DataGridProps) {
     this.instance?.on('optionChanged', this.instanceOptionChangedHandler.bind(this));
   }
 
-  instanceOptionChangedHandler(e: any): void {
-    const isValueCorrect = e.value === e.component.option(e.fullName); // T867777
+  instanceOptionChangedHandler(e: OptionChangedEvent): void {
+    try {
+      this.isUpdate = true;
```
Comment (vconst) [inline js/renovation/ui/grids/data_grid/data_grid.tsx]: isUpdate -> isTwoWayPropUpdating

File: js/renovation/component_wrapper/utils.ts [hunk 6]
```diff
@@ -29,3 +30,31 @@ export const removeDifferentElements = ($children, $newChildren): void => {
...
+          newArray[index] = { ...currentPropsValue[index] };
+        }
+      }
+      return;
+    }
+  }
+  if (isObject(currentPropsValue) && isPlainObject(currentPropsValue)) {
```
Comment (vconst) [inline js/renovation/component_wrapper/utils.ts]: Remove isObject from here.

**Review by pavelgruba — APPROVED** [review]

**Review by vconst — APPROVED** [review]

**Pull request merged**